### PR TITLE
Creating Ontology Mapper

### PR DESCRIPTION
**Pull request by joebeernink**
Creating Initial Version of Ontology Mapper

### CODE REVIEW

File: OntologyMapper/test/OntologyMapper.Test/OntologyMappingTests.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            ontologyMapping.FillProperties.Add(new FillProperty { OutputDtmiFilter = "*", OutputPropertyName = "name", InputPropertyNames = "name description" });
+
+            ontologyMapping.InterfaceRemaps.Add(new DtmiRemap { InputDtmi = "inputDtmi1", OutputDtmi = "outputDtmi1" });
+            ontologyMapping.InterfaceRemaps.Add(new DtmiRemap { InputDtmi = "inputDtmi2", OutputDtmi = "outputDtmi2" });
+
+            ontologyMapping.RelationshipRemaps.Add(new RelationshipRemap { InputRelationship = "isA", OutputRelationship = "wasA" });
+
+            var serialized = JsonConvert.SerializeObject(ontologyMapping, Formatting.Indented);
+            output.WriteLine(serialized);
+        }
```
Comment (Tyler-Angell) [inline OntologyMapper/test/OntologyMapper.Test/OntologyMappingTests.cs:43]: There are no asserts in this test. Please add some basic assertion. Otherwise its a tool, not a test.

File: OntologyMapper/test/OntologyMapper.Test/OntologyMapper.Test.csproj [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+		<PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="6.0.1" />
+		<PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.11.0"  />
+		<PackageReference Include="MSTest.TestAdapter" Version="2.2.7" />
+		<PackageReference Include="MSTest.TestFramework" Version="2.2.7" />
+		<PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
+		<PackageReference Include="xunit.extensibility.core" Version="2.4.1" />
+	</ItemGroup>
+
+	<ItemGroup>
+		<ProjectReference Include="..\..\src\OntologyMapper\OntologyMapper.csproj" />
```
Comment (Tyler-Angell) [inline OntologyMapper/test/OntologyMapper.Test/OntologyMapper.Test.csproj:26]: The OntologyMapper and OntologyMapper.Test folders are redundant under the src and test folders

OntologyMapper
- src
- - OntologyMapper.csproj
- test
-- OntologyMapper.Tests.csproj

File: README.md [hunk 3]
```diff
@@ -20,3 +20,6 @@ Once you have uploaded these models into your Azure Digital Twins instance, you
 ## Validating the models
 The DTDL RealEstateCore models in this repo have already been validated. You don't have to validate them with the DTDL parser unless you change them. If you have extended the models or made changes, it's recommended to validate the models as described by this article: [Validate models](https://docs.microsoft.com/en-us/azure/digital-twins/concepts-convert-models#validate-and-upload-dtdl-models).
 
+## Converting from one ontology / version to another
+The OntologyMapper is an assembly that is used to support conversion from one DTDL ontology to another DTDL ontology. It is used in conjuction with software that loads an topology from a source and inserts to a target.
```
Comment (Tyler-Angell) [inline README.md:24]: When it says `with software` is that referring to something I as an end user would write or what software is this referring to?

File: OntologyMapper/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+## Motivation and purpose
+
+The OntologyMapper allows developers to define conversion processes from one DTDL-based Ontology to a different DTDL-based Ontology.
+
+This may be needed in cases where cloning a topology which uses a propriety ontology, and converting to an open standard ontology such as Brick or RealEstateCore.
+
+These conversion rules are defined in a series of Mappings files here in the repo, which are currently embedded in the assembly to make versioning easier.
+
+The key is that these files describe exceptions only. If the source DTMI and Target DTMI are identical, nothing needs to appear for the mapping between the two. Thus, if an source and target use the 
+same ontology and version, the mapping file will have very little content.
```
Comment (Tyler-Angell) [inline OntologyMapper/README.md:12]: I think there is an unintended newline here

File: OntologyMapper/README.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+The Mappings folder is a hierarchy which goes from the Input Ontology, format, and version, to the output ontology. A version of the file should exist for each combination of versions of the source to target ontology for each combination actually in use. 
+
+The name of the file is currently important, as it is used as part of the start up configuration for processes using these files, so changing the name (or having duplicate names anywhere in the assembly), will cause runtime errors.
+
+## Mapping Classes
+
+### OntologyMapping
+
+This class defines the structure of the files stored in the Mappings folder for deserialization. The structure is defined [here] (./src/OntologyMapping.cs)
```
Comment (Tyler-Angell) [inline OntologyMapper/README.md:24]: This is mentioned in another comment, but the file should exist here, but it currently doesn't.

File: OntologyMapper/README.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+### OntologyMappingManager
+
+This class defines an implementation of the IOntologyMappingManager, and provides simplified calls for clients to make to use the OntologyMapping
```
Comment (Tyler-Angell) [inline OntologyMapper/README.md:28]: Can you please provide a short annotated example of how one crafts a mapping file? There are several sections in one of those and I'm not quite following how to get from A to B when it comes to properties

File: OntologyMapper/src/OntologyMapper/OntologyMapping.cs [hunk 7]
```diff
@@ -0,0 +1,167 @@
+﻿// -----------------------------------------------------------------------
+// <copyright file="OntologyMapping.cs" company="Microsoft">
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// </copyright>
+// -----------------------------------------------------------------------
+
+namespace Microsoft.SmartFacilities.OntologyMapper
```
Comment (Tyler-Angell) [inline OntologyMapper/src/OntologyMapper/OntologyMapping.cs:7]: This is starting off with SmartFacilities, but the OntologyMapper could be used for all sorts of verticals. Since its DTDL -> DTDL I wouldn't expect there shouldn't be any domain specific logic.

File: OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    using System.Reflection;
+
+    public class ResourceFileOntologyMappingLoader : IOntologyMappingLoader
+    {
+        private readonly ILogger logger;
+        private readonly string resourcePath = string.Empty;
+
+        public ResourceFileOntologyMappingLoader(ILogger logger, string resourcePath)
+        {
+            if (string.IsNullOrEmpty(resourcePath))
```
Comment (Tyler-Angell) [inline OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs:20]: IsNullOrWhitespace?

File: OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+                        string result = reader.ReadToEnd();
+                        var mappings = JsonConvert.DeserializeObject<OntologyMapping>(result);
+
+                        if (mappings != null)
+                        {
+                            return mappings;
+                        }
+                        else
+                        {
+                            var error = $"Mappings file '{resourcePath}' is empty or poorly formed.";
```
Comment (Tyler-Angell) [inline OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs:51]: Since these mappings are coming from an embedded resource we should probably have a test to validated them before being checked in, similar to the environment definition files from out last team so bad ones dont get checked in.

File: OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+                            throw new Exception(error);
```
Comment (Tyler-Angell) [inline OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs:52]: Should we create a custom Exception for this?

File: OntologyMapper/test/OntologyMapper.Test/OntologyMappingTests.cs (hunk 1 above)
Comment (joebeernink) [reply OntologyMapper/test/OntologyMapper.Test/OntologyMappingTests.cs]: Removed file. No longer needed

File: OntologyMapper/test/OntologyMapper.Test/OntologyMapper.Test.csproj (hunk 2 above)
Comment (joebeernink) [reply OntologyMapper/test/OntologyMapper.Test/OntologyMapper.Test.csproj]: Reorganized files

File: README.md (hunk 3 above)
Comment (joebeernink) [reply README.md]: Removed line until we determine shipping model for the rest of the product

File: OntologyMapper/README.md (hunk 5 above)
Comment (joebeernink) [reply OntologyMapper/README.md:23]: Should now

File: OntologyMapper/README.md (hunk 6 above)
Comment (joebeernink) [reply OntologyMapper/README.md:27]: I left a lot of comments in the Mapping file itself so that the code is documented. Not sure if that is sufficient

File: OntologyMapper/src/OntologyMapper/OntologyMapping.cs (hunk 7 above)
Comment (joebeernink) [reply OntologyMapper/src/OntologyMapper/OntologyMapping.cs]: Not sure what namespace to put this in at this point. And what are the Microsoft guideline for it. We will likely move it before creating the nuget

File: OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs (hunk 9 above)
Comment (joebeernink) [reply OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs]: In progress

Comment (joebeernink) [reply OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs]: The issue is also having a copy of the valid model in the repo, which would be even more problematic

File: OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs (hunk 10 above)
Comment (joebeernink) [reply OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs]: Done

File: OntologyMapper/src/OntologyMapper/OntologyMappingManager.cs [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        public OntologyMappingManager(IOntologyMappingLoader mappingLoader)
+        {
+            ontologyMapping = mappingLoader.LoadOntologyMapping();
+        }
+
+        public bool TryGetInterfaceRemapDtmi(Dtmi inputDtmi, out DtmiRemap? dtmiRemap)
+        {
+            dtmiRemap = ontologyMapping.InterfaceRemaps.FirstOrDefault(r => r.InputDtmi == inputDtmi.ToString() && !r.IsIgnored);
+
+            if (dtmiRemap != null)
```
Comment (Tyler-Angell) [inline OntologyMapper/src/OntologyMapper/OntologyMappingManager.cs]: Since its FirstOrDefault wouldn't the value not be null?

File: OntologyMapper/src/OntologyMapper/OntologyMappingManager.cs [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+            {
+                return true;
+            }
+
+            return false;
```
Comment (Tyler-Angell) [inline OntologyMapper/src/OntologyMapper/OntologyMappingManager.cs]: This seems backwards? Shouldn't successfully retrieving it return true?

File: OntologyMapper/src/OntologyMapper.csproj [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    <Description>Support for converting from one DTDL-based Ontology to a different DTDL Based Ontology</Description>
+    <AssemblyVersion>0.1.0</AssemblyVersion>
+    <RepositoryUrl>https://github.com/Azure/opendigitaltwins-tools</RepositoryUrl>
+    <Copyright>Copyright (c) Microsoft Corporation.  All rights reserved.</Copyright>
+    <AssemblyName>Microsoft.SmartFacilities.OntologyMapper</AssemblyName>
+    <RootNamespace>Microsoft.SmartFacilities.OntologyMapper</RootNamespace>
+  </PropertyGroup>
+	
+  <ItemGroup>
+    <EmbeddedResource Include="Mappings\Mapped\Json\v0\BrickRec\mapped_json_v0_dtdlv2_Brick_1_3-REC_4_0.json" />
```
Comment (Tyler-Angell) [inline OntologyMapper/src/OntologyMapper.csproj:19]: Would it make sense to add these files as a wildcard? Seems like it would be easy for newcomers to miss

File: OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs (hunk 9 above)
Comment (Tyler-Angell) [reply OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs]: You could validate that the mapping files all follow the schema so that even if exact values are wrong, at least the file would load? Could be a compromise between making sure the mappings are valid and the making sure the mappings aren't poorly formed

File: OntologyMapper/README.md (hunk 6 above)
Comment (Tyler-Angell) [reply OntologyMapper/README.md:27]: If you're learning a new tool do you want to dive through the code and figure out how it all works and hope there are comments in there? Or would you rather have some sort of guidance? I know initially I dont care about how "it" works, I just want to know the input and how to use it with a couple of examples. If it does what I want it to and its relevant than I'll dig deeper into what it can d

Maybe at least point to the file where all the code is documented so people dont have to dig through every file to find it. But I wouldn't really expect to find explanations for json in a cs file.

File: OntologyMapper/README.md [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+These conversion rules are defined in a series of Mappings files here in the repo, which are currently embedded in the assembly to make versioning easier.
+
+The key is that these files describe exceptions only. If the source DTMI and Target DTMI are identical, nothing needs to appear for the mapping between the two. Thus, if an source and target use the same ontology and version, the mapping file will have very little content.
+
+## Mappings Folder
+
+The Mappings folder is a hierarchy which goes from the Input Ontology, format, and version, to the output ontology. A version of the file should exist for each combination of versions of the source to target ontology for each combination actually in use. 
+
+The name of the file is currently important, as it is used as part of the start up configuration for processes using these files, so changing the name (or having duplicate names anywhere in the assembly), will cause runtime errors.
+
```
Comment (Tyler-Angell) [inline OntologyMapper/README.md:18]: I'd suggest adding a visual sample of what you're explaining

```shell
│
├───OntologyMapper
│   ├───src
│   │   ├───Mappings
│   │   │   ├───<Input Ontologies Name>
│   │   │   │   ├───<Format>
│   │   │   │   │   ├───<Mapping Version>
│   │   │   │   │   │   ├───<Output Ontologies Name>
│   │   │   │   │   │   │      <unique name>.json```

File: OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs (hunk 9 above)
Comment (Tyler-Angell) [reply OntologyMapper/src/OntologyMapper/ResourceFileOntologyMappingLoader.cs]: Since the mapping jsons are also a key, we'd probably need assurance that they are project unique and someone didn't add two versions of the same name despite being in different folders

File: OntologyMapper/src/OntologyMapper/OntologyMappingManager.cs (hunk 11 above)
Comment (joebeernink) [reply OntologyMapper/src/OntologyMapper/OntologyMappingManager.cs]: Default is null

File: OntologyMapper/src/OntologyMapper.csproj [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+    <EmbeddedResource Include="Mappings\Mapped\Json\v0\BrickRec\mapped_json_v0_dtdlv3_Brick_1_3-REC_4_0.json" />
```
Comment (Tyler-Angell) [inline OntologyMapper/src/OntologyMapper.csproj:20]: As the number of contributors grows, so will the size of this library. I can see adding an un-embedded version of the IOntologyMappingLoader for those to add/test their own mappings if their DTDLs are sensitive, but does it make sense Mapped to be included in the core implementation of the mapping library so that everyone who uses it is required to add it to their project? Some edge case scenarios may be sensitive to project size.

File: OntologyMapper/test/OntologyMappingManagerTests.cs [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+        }
+
+        //private IReadOnlyDictionary<Dtmi, DTEntityInfo> GetTargetObjectModel()
+        //{
+        //    var targetObjectModel = new Dictionary<Dtmi, DTEntityInfo>();
+
+        //    targetObjectModel.Add(new Dtmi("dtmi:mapped:core:AblutionsRoom;1"), new DTEntityInfo());
+
+        //    return targetObjectModel;
+        //}
```
Comment (Tyler-Angell) [inline OntologyMapper/test/OntologyMappingManagerTests.cs:213]: Is this obsolete?

File: OntologyMapper/test/OntologyMappingManagerTests.cs [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+            Assert.False(result);
+            Assert.NotNull(propertyFill);
+            Assert.Empty(propertyFill);
+        }
+
+        private OntologyMapping GetOntologyMapping()
+        {
+            var ontologyMapping = new OntologyMapping();
+
+            ontologyMapping.Header.InputOntologies.Add(new Ontology { DtdlVersion = "v0", Name = "mapped", Version = "1.0" });
```
Comment (Tyler-Angell) [inline OntologyMapper/test/OntologyMappingManagerTests.cs:190]: For clarity since this is test code, it might be clearer to make these not real values

File: OntologyMapper/src/OntologyMapper/OntologyMappingManager.cs (hunk 12 above)
Comment (joebeernink) [reply OntologyMapper/src/OntologyMapper/OntologyMappingManager.cs]: Looks ok to me... tests pass

File: OntologyMapper/test/OntologyMappingManagerTests.cs (hunk 16 above)
Comment (joebeernink) [reply OntologyMapper/test/OntologyMappingManagerTests.cs:213]: Haven't finished writing this test yet

File: OntologyMapper/src/OntologyMappingManager.cs [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+        private OntologyMapping ontologyMapping;
+
+        public OntologyMappingManager(IOntologyMappingLoader mappingLoader)
+        {
+            ontologyMapping = mappingLoader.LoadOntologyMapping();
+        }
+
+        public bool TryGetInterfaceRemapDtmi(Dtmi inputDtmi, out DtmiRemap? dtmiRemap)
+        {
+            dtmiRemap = ontologyMapping.InterfaceRemaps.FirstOrDefault(r => r.InputDtmi == inputDtmi.ToString() && !r.IsIgnored);
```
Comment (Tyler-Angell) [inline OntologyMapper/src/OntologyMappingManager.cs:23]: For my own learning: dtmiRemap is the out variable, whereas the rest in this file all set the firstOrDefault to result first before reassigning to the out variable. Is there a difference?

File: OntologyMapper/src/Mappings/Mapped/Json/v0/BrickRec/mapped_json_v0_dtdlv3_Brick_1_3-REC_4_0.json [hunk 19]
```diff
@@ -0,0 +1,5494 @@
...
+      "OutputPropertyName": "externalIds",
+      "InputPropertyName": "mappingKey",
+      "IsOutputPropertyCollection": true
+    }
+  ],
+  "FillProperties": [
+    {
+      "OutputDtmiFilter": "*",
+      "OutputPropertyName": "name",
+      "InputPropertyNames": "name description"
```
Comment (Tyler-Angell) [inline OntologyMapper/src/Mappings/Mapped/Json/v0/BrickRec/mapped_json_v0_dtdlv3_Brick_1_3-REC_4_0.json:5491]: How can more than one Input fill a single Output? Should this be an array?

File: OntologyMapper/src/OntologyMapper/OntologyMappingManager.cs (hunk 12 above)
Comment (Tyler-Angell) [reply OntologyMapper/src/OntologyMapper/OntologyMappingManager.cs]: Whoops, my bad. I was reading `dtmiRemap == null`

File: OntologyMapper/src/OntologyMappingManager.cs (hunk 18 above)
Comment (joebeernink) [reply OntologyMapper/src/OntologyMappingManager.cs]: Cleaned up to make more consistent

File: OntologyMapper/src/Mappings/Mapped/Json/v0/BrickRec/mapped_json_v0_dtdlv3_Brick_1_3-REC_4_0.json (hunk 19 above)
Comment (joebeernink) [reply OntologyMapper/src/Mappings/Mapped/Json/v0/BrickRec/mapped_json_v0_dtdlv3_Brick_1_3-REC_4_0.json:5491]: Converted to an array

File: OntologyMapper/test/OntologyMappingManagerTests.cs (hunk 16 above)
Comment (joebeernink) [reply OntologyMapper/test/OntologyMappingManagerTests.cs]: Done

File: OntologyMapper/README.md [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+##### Sample FillProperties
+
+``` json
+
+"FillProperties": [
+    {
+      "OutputDtmiFilter": "*",
+      "OutputPropertyName": "name",
+      "InputPropertyNames": "name description"
```
Comment (Tyler-Angell) [inline OntologyMapper/README.md:147]: [ "name, description" ]

**Review by Tyler-Angell — APPROVED** [review]
